docs/nnx/bridge_guide.md

```diff
@@ -0,0 +1,544 @@
+
+**Note**: 
+
+This guide is about glueing Linen and NNX modules. To migrate an existing Linen module to NNX, check out the [Migrate from Linen to NNX](https://flax.readthedocs.io/en/latest/nnx/haiku_linen_vs_nnx.html) guide. 
+
+And all built-in Linen layers should have equivalent NNX versions! Check out the list of [Built-in NNX layers](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/nn/index.html).
+
+
+```python
+import os
+os.environ["XLA_FLAGS"] = '--xla_force_host_platform_device_count=8'
+
+from flax import nnx
+from flax import linen as nn
+from flax.nnx import bridge
+import jax
+from jax import numpy as jnp
+from jax.experimental import mesh_utils
+from typing import *
+```
+
+## Submodule is all you need
+
+A Flax model is always a tree of modules - either old Linen modules (`flax.linen.Module`, usually written as `nn.Module`) or NNX modules (`nnx.Module`). 
+
+An `nnx.bridge` wrapper glues the two types together, in both ways:
+
+* `nnx.bridge.ToNNX`: Convert a Linen module to NNX, so that it can be a submodule of another NNX module, or stand alone to be trained in NNX-style training loops.
+* `nnx.bridge.ToLinen`: Vice versa, convert a NNX module to Linen.
+
+This means you can move in either top-down or bottom-up behavior: convert the whole Linen module to NNX, then gradually move down, or convert all the lower level modules to NNX then move up.
+
+
+## The Basics
+
+There are two fundamental difference between Linen and NNX modules:
+
+* **Stateless vs. stateful**: Linen module instances are stateless: variables are returned from a purely functional `.init()` call and managed separately. NNX modules, however, owns its variables as instance attributes.
+
+* **Lazy vs. eager**: Linen modules only allocate space to create variables when they actually see their input. Whereas NNX module instances create their variables the moment they are instantiated, without seeing a sample input.
+
+With that in mind, let's look at how the `nnx.bridge` wrappers tackle the differences.
+
+### Linen -> NNX
+
+Since Linen modules may require an input to create variables, we semi-formally supported lazy initialization in the NNX modules converted from Linen. The Linen variables are created when you give it a sample input.
+
+For you, it's calling `nnx.bridge.lazy_init()` where you call `module.init()` in Linen code.
+
+
+```python
+class LinenDot(nn.Module):
+  out_dim: int
+  w_init: Callable[..., Any] = nn.initializers.lecun_normal()
+  @nn.compact
+  def __call__(self, x):
+    # Linen might need the input shape to create the weight!
+    w = self.param('w', self.w_init, (x.shape[-1], self.out_dim))
+    return x @ w
+
+x = jax.random.normal(jax.random.key(42), (4, 32))
+model = bridge.ToNNX(LinenDot(64), rngs=nnx.Rngs(0))  # => `model = LinenDot(64)` in Linen
```

Comment:
It would be interesting to  do an `nnx.display` here after construction to visualize the uninitialized structure, and in a new cell call the rest and visualize the module again after its initialized.

Reply:
Good idea! Done.